modules/gdscript/gdscript_analyzer.cpp

```diff
@@ -1584,17 +1584,34 @@ void GDScriptAnalyzer::resolve_assignable(GDScriptParser::AssignableNode *p_assi
 				if (is_variable) {
 					static_cast<GDScriptParser::VariableNode *>(p_assignable)->use_conversion_assign = true;
 				}
-			} else if (!is_type_compatible(specified_type, initializer_type, true, p_assignable->initializer)) {
-				if (is_variable && is_type_compatible(initializer_type, specified_type, true, p_assignable->initializer)) {
+			} else if (!is_type_compatible(initializer_type, specified_type, true)) {
+				push_error(vformat(R"(Cannot assign a value of type %s to %s "%s" with specified type %s.)", initializer_type.to_string(), p_kind, p_assignable->identifier->name, specified_type.to_string()), p_assignable->initializer);
+			} else {
+#ifdef DEBUG_ENABLED
+				if (specified_type.builtin_type == Variant::INT && initializer_type.builtin_type == Variant::FLOAT) {
+					parser->push_warning(p_assignable->initializer, GDScriptWarning::NARROWING_CONVERSION);
+				}
+#endif
+
+				Variant::Type initializer_builtin_type = type_from_metatype(initializer_type).builtin_type;
+				Variant::Type specified_builtin_type = type_from_metatype(specified_type).builtin_type;
+
+				// TODO: Handle arrays correctly.
+				if (p_assignable->initializer->is_constant && initializer_builtin_type != specified_builtin_type && Variant::can_convert_strict(initializer_builtin_type, specified_builtin_type) && p_assignable->initializer->type != GDScriptParser::Node::ARRAY) {
```

Comment:
We need to think about what conversions should be allowed. In the related issue, the problem is that `const X: int = 1.0` doesn't work. But `Variant::can_convert_strict` allows too much, like `const X: int = false`. At a minimum, a warning is needed, but some analogue of `is_type_compatible` is probably needed.

Reply:
I checked, it's already broken in beta 11. The first part has room for discussion, but the second is a bug. I'll try to fix it in this PR as I've added a check for constant initializers.

```gdscript
@tool
extends EditorScript

func _run() -> void:
    var a: int = false
    var b: float = true
    var c: bool = 1.0
    prints(var_to_str(a)) # 0
    prints(var_to_str(b)) # 1.0
    prints(var_to_str(c)) # true
    
    # **Runtime errors**, not complile time.
    var d: int = null # Trying to assign value of type 'Nil' to a variable of type 'int'.
    var e: float = null # Trying to assign value of type 'Nil' to a variable of type 'float'.
    var f: bool = null # Trying to assign value of type 'Nil' to a variable of type 'bool'.
```

Reply:
> ```gdscript
>     # **Runtime errors**, not complile time.
>     var d: int = null # Trying to assign value of type 'Nil' to a variable of type 'int'.
>     var e: float = null # Trying to assign value of type 'Nil' to a variable of type 'float'.
>     var f: bool = null # Trying to assign value of type 'Nil' to a variable of type 'bool'.
> ```

Fixed.